### PR TITLE
Pydantic improvements

### PR DESCRIPTION
**Pull request by ZergLev**
# Description:

- TLDR: All classes in /chatsky/pipeline/ are rewritten in Pydantic, some changes to class hierarchy and the classes themselves. Changed tutorials with regards to new class signatures.
- Classes `PipelineComponent`, `Pipeline` and `ComponentExtraHandler` are now subclasses of Pydantic's `BaseModel`, massively reducing code bloat in their initialization and that of their children (such as `Service`).
- Also, classes `Service`, `ServiceGroup` and `Actor` are now subclasses of `PipelineComponent`. No functionality is removed in `ServiceGroup`. All tutorials and tests pass like before, with a few changes.
- Reasoning for changing class hierarchy: Before this, an instance of `Actor` would be called from within a `Service` with a special flag, which seemed like a workaround, and now it can be called directly as an instance of `PipelineComponent`, instead of heavily relying on the `Service` class. You could say that this slightly reduces the code's complexity. In fact, it also binds `Actor` to `PipelineComponent`, instead of it being independent, making code look more cohesive.
- TypeBuilders have been removed, because Pydantic can already deal with them effectively, giving type hints for IDE like before. Different input types are still handled like before, except nested `Service` declaration is removed. (`Service` class doesn't accept `Service` as it's handler anymore) Same with nested `ComponentExtraHandler`s.
- For some reason Extra Handlers in `Service`s would work even if the Service itself was turned off via `start_condition`. That was a bug. Now this condition is checked in `PipelineComponent` instead. I assume `Actor` must always have it as True.

# Checklist

- [x] Removed `from_script()` and other redundant methods.
- [x] Changed `components` into `pre-services` and `post-services` throughout tutorials. (at `Pipeline` initialization)
- [x] Got tests to pass.
- [x] I have performed a self-review of the changes
- [x] Updated Pipeline's API and tutorials

# To Consider

- Add more tests if necessary
- Update the remaining API reference / tutorials / guides
- Search for more references to changed entities in the codebase

### CODE REVIEW

File: chatsky/pipeline/service/group.py [hunk 1]
```diff
@@ -44,64 +41,48 @@ class ServiceGroup(PipelineComponent):
...
-            super(ServiceGroup, self).__init__(
-                before_handler, after_handler, timeout, asynchronous, calc_async, start_condition, name
-            )
+    components: List[
+        Union[
+            Actor,
+            Service,
+            ServiceGroup,
+        ]
+    ]
```
Comment (RLKRo) [inline chatsky/pipeline/service/group.py:63]: Why not `List[PipelineComponent]`?

File: chatsky/pipeline/service/group.py [hunk 2]
```diff
@@ -35,7 +32,7 @@
     from chatsky.pipeline.pipeline.pipeline import Pipeline
 
 
-class ServiceGroup(PipelineComponent):
+class ServiceGroup(PipelineComponent, extra="forbid", arbitrary_types_allowed=True):
```
Comment (RLKRo) [inline chatsky/pipeline/service/group.py:35]: Comments from extra.py apply here and in service.py as well.

File: chatsky/pipeline/service/service.py [hunk 3]
```diff
@@ -127,66 +94,15 @@ async def _run_handler(self, ctx: Context, pipeline: Pipeline) -> None:
...
     def info_dict(self) -> dict:
         """
         See `Component.info_dict` property.
         Adds `handler` key to base info dictionary.
         """
         representation = super(Service, self).info_dict
-        if isinstance(self.handler, str) and self.handler == "ACTOR":
-            service_representation = "Instance of Actor"
-        elif callable(self.handler):
+        # Need to carefully remove this
```
Comment (RLKRo) [inline chatsky/pipeline/service/service.py:104]: Could you elaborate?

File: tests/pipeline/test_validation.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        pass
+
+
+# Could make a test for returning an awaitable from a ServiceFunction, ExtraHandlerFunction
+class TestServiceValidation:
+    def test_model_validator(self):
+        with pytest.raises(ValidationError) as e:
+            # Can't pass a list to handler, it has to be a single function
+            Service(handler=[UserFunctionSamples.correct_service_function_2])
+            assert e
```
Comment (RLKRo) [inline tests/pipeline/test_validation.py:41]: No need to `assert e`. Can simply do:
```python
with pytest.raises(ValidationError):
    # code
```

File: tests/pipeline/test_validation.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        with pytest.raises(ValidationError) as e:
+            # 'handler' is a mandatory field
+            Service(before_handler=UserFunctionSamples.correct_service_function_2)
```
Comment (RLKRo) [inline tests/pipeline/test_validation.py:44]: Tests don't need to be so thorough.
In this case we rely on Pydantic and developers/reviewers (to ensure correct field definition) to make sure that a service cannot be created without passing a handler.

File: tutorials/pipeline/3_pipeline_dict_with_services_basic.py [hunk 6]
```diff
@@ -31,18 +31,18 @@
 
 # %% [markdown]
 """
-When Pipeline is created using `from_dict` method,
+When Pipeline is created using Pydantic's `model_validate` method,
```
Comment (RLKRo) [inline tutorials/pipeline/3_pipeline_dict_with_services_basic.py:34]: Since Pipeline no longer has the `from_script` method (and there's no difference between using `__init__` or `model_validate`), I think these tutorials should be merged into one that just goes over Pipeline fields and mentions that it can be initialized from a dictionary.

File: chatsky/pipeline/service/service.py (hunk 3 above)
Comment (ZergLev) [reply chatsky/pipeline/service/service.py:104]: I think I meant that since ACTOR is no longer a valid `Service`, this should be changed. And because there is really only one type of handler now (Callable), the whole "if-else" block could be removed and the f-string inserted into the dict.update(). It's unlikely the handler will be "unknown", right?

File: chatsky/pipeline/pipeline/pipeline.py [hunk 7]
```diff
@@ -83,61 +82,62 @@ class Pipeline:
...
+    before_handler: ComponentExtraHandler = Field(default_factory=list)
+    after_handler: ComponentExtraHandler = Field(default_factory=list)
+    timeout: Optional[float] = None
+    optimization_warnings: bool = False
+    parallelize_processing: bool = False
+    actor: Optional[Actor] = None
+    _services_pipeline: Optional[ServiceGroup] = None
+    _clean_turn_cache: Optional[bool]
+
+    def _create_actor(self) -> Actor:
```
[the remark this answers is not in the document]
Comment (ZergLev) [reply chatsky/pipeline/pipeline/pipeline.py:105]: I browsed the net, and in case someone directly changes a constructor parameter in Pipeline, `computed_field` will start recomputing, breaking the program. I'm not sure what happens if the changed parameter is unrelated to said `computed field`, like when a Pipeline's flag is changed after it's creation.
In case the change is indirect, like "pre_services.components[0].set_state(FINISHED)" the program probably won't start recomputing `_services_pipeline`, right? I haven't found much on the internet about this.

File: chatsky/pipeline/service/service.py (hunk 3 above)
Comment (RLKRo) [reply chatsky/pipeline/service/service.py:104]: Got it.
I think `PipelineComponent` should have a property `computed_name` that returns names produced by `rename_component_incrementing`.
So for service it would be
```
if isfunction(self.handler);
  return self.handler.__name__
else:
  return self.handler.__class__.__name__
```
Then this property can be called in `rename_component_incrementing` and here (`info_dict`).

File: chatsky/pipeline/pipeline/actor.py [hunk 8]
```diff
@@ -67,36 +84,45 @@ class Actor:
...
-        fallback_label: Optional[NodeLabel2Type] = None,
-        label_priority: float = 1.0,
-        condition_handler: Optional[Callable] = None,
-        handlers: Optional[Dict[ActorStage, List[Callable]]] = None,
-    ):
-        self.script = script if isinstance(script, Script) else Script(script=script)
-        self.label_priority = label_priority
-
-        self.start_label = normalize_label(start_label)
+    script: Union[Script, dict]
```
[the remark this answers is not in the document]
Comment (ZergLev) [reply chatsky/pipeline/pipeline/actor.py]: I just tried that and Dict is not cast to `Script` automatically. Script doesn't have a model_validator like that right now. I could make one, it's easy. Actor was already doing that before.

[the remark this answers is not in the document]
Comment (RLKRo) [reply chatsky/pipeline/pipeline/actor.py]: Oh, right. This is an old version of `Script`. In core rework it is changed and can be validated from dict.
No need to change it now, I'll update it with the core rework.

File: chatsky/pipeline/service/extra.py [hunk 9]
```diff
@@ -30,53 +29,43 @@
...
+    functions: List[ExtraHandlerFunction] = Field(default_factory=list)
+    stage: ClassVar[ExtraHandlerType] = ExtraHandlerType.UNDEFINED
+    timeout: Optional[float] = None
+    requested_async_flag: Optional[bool] = None
+
+    @model_validator(mode="before")
+    @classmethod
+    # Here Script class has "@validate_call". Is it needed here?
+    def functions_constructor(cls, data: Any):
+        if not isinstance(data, dict):
```
[the remark this answers is not in the document]
Comment (ZergLev) [reply chatsky/pipeline/service/extra.py]: True. Though, when I tried to fix it, I noticed that `Extra Handlers` have a `__call__` method, so they are `Callable`s, with the right signatures too. Technically, we could forbid `Extra Handlers` to be passed as `Callable`s there, but I actually don't see any problems with that, it shouldn't break the program.

File: chatsky/pipeline/pipeline/component.py [hunk 10]
```diff
@@ -60,50 +61,23 @@ class PipelineComponent(abc.ABC):
...
-        """
-        self.path = path
-        """
-        Dot-separated path to component (is universally unique).
-        This attribute is set in :py:func:`~chatsky.pipeline.pipeline.utils.finalize_service_group`.
-        """
-
-        self.before_handler = BeforeHandler([] if before_handler is None else before_handler)
-        self.after_handler = AfterHandler([] if after_handler is None else after_handler)
+    before_handler: BeforeHandler = Field(default_factory=BeforeHandler)
```
[the remark this answers is not in the document]
Comment (ZergLev) [reply chatsky/pipeline/pipeline/component.py:64]: Did that for PipelineComponent and ComponentExtraHandler. What about Service and ServiceGroup? Will there only be 'handler' docstrings for Service? Or will it carry over the docstrings from it's parent class, PipelineComponent? Also, should I make the same changes for Actor and Pipeline too?

**Comment by ZergLev** [comment]
'inherited-members' in autodoc_default_options allows all classes to be inherited, except those explicitly defined as it's arguments. Which is the exact opposite of what I need. If I could just inherit docstrings from `PipelineComponent`, there'd be no problem.
Basically, I don't see any good solution for this using this method. I remind, 'handler' from `Service` and 'components' from `ServiceGroup` are undocumented in the relevant API, but they are in the constructor parameters, which are right next to that in the API reference. I could add docs for them, but then they will be duplicate, two docs for the same field on the same page. Alternatively, I could remove constructor parameters and type all those parameters in the same file, essentially duplicating the docs. I'll have to let it hang for now.

File: tutorials/pipeline/3_pipeline_dict_with_services_basic.py (hunk 6 above)
Comment (ZergLev) [reply tutorials/pipeline/3_pipeline_dict_with_services_basic.py]: These two tutorials both only had `from_dict`. I think the basic version exists for rookie developers, because they may get caught up reading Pipeline's signature. Sure, expreienced devs will glance at it and keep all that in mind, but newbies may have difficulty processing that. For now I made the merged tutorial in the basic tutorial.

File: chatsky/pipeline/pipeline/component.py [hunk 11]
```diff
@@ -40,40 +40,56 @@ class PipelineComponent(abc.ABC, BaseModel, extra="forbid", arbitrary_types_allo
...
     path: Optional[str] = None
+    """
+    Separated by dots path to component, is universally unique.
+    """
 
     @model_validator(mode="after")
-    def pipeline_component_validator(self):
-        if self.name is not None and (self.name == "" or "." in self.name):
-            raise Exception(f"User defined service name shouldn't be blank or contain '.' (service: {self.name})!")
+    def __pipeline_component_validator(self):
```
Comment (RLKRo) [inline chatsky/pipeline/pipeline/component.py:87]: Add doc to validators: need to document when they raise and what data they add support for (i.e. validation from `Callable` for services).

File: chatsky/pipeline/pipeline/component.py [hunk 12]
```diff
@@ -134,9 +150,35 @@ async def run_extra_handler(self, stage: ExtraHandlerType, ctx: Context, pipelin
...
+        """
+        Method for running this component. It can be an Actor, Service or ServiceGroup.
+        It has to be defined in the child classes,
+        which is done in each of the default PipelineComponents.
+        Service 'handler' has three possible signatures. These possible signatures are:
+
+        - (ctx: Context) - accepts current dialog context only.
+        - (ctx: Context, pipeline: Pipeline) - accepts context and current pipeline.
+        - | (ctx: Context, pipeline: Pipeline, info: ServiceRuntimeInfo) - accepts context,
+              pipeline and service runtime info dictionary.
```
Comment (RLKRo) [inline chatsky/pipeline/pipeline/component.py:163]: Not relevant to PipelineComponent.

File: chatsky/pipeline/pipeline/component.py [hunk 13]
```diff
@@ -40,40 +40,56 @@ class PipelineComponent(abc.ABC, BaseModel, extra="forbid", arbitrary_types_allo
...
+    """
+    Whether the component can be asynchronous or not.
+
+    1) for :py:class:`~.pipeline.service.service.Service`: whether its `handler` is asynchronous or not,
+    2) for :py:class:`~.pipeline.service.group.ServiceGroup`: whether all its `services` are asynchronous or not.
+
+    """
     start_condition: StartConditionCheckerFunction = Field(default=always_start_condition)
+    """
+    StartConditionCheckerFunction that is invoked before each component execution;
```
Comment (RLKRo) [inline chatsky/pipeline/pipeline/component.py:73]: Could you go over pipeline docs and fix issues like missing links?

File: chatsky/pipeline/pipeline/component.py [hunk 14]
```diff
@@ -134,9 +150,35 @@ async def run_extra_handler(self, stage: ExtraHandlerType, ctx: Context, pipelin
             logger.warning(f"{type(self).__name__} '{self.name}' {extra_handler.stage} extra handler timed out!")
 
     @abc.abstractmethod
-    async def run_component(self, ctx: Context, pipeline: Pipeline) -> None:
+    async def run_component(self, ctx: Context, pipeline: Pipeline) -> Optional[ComponentExecutionState]:
+        """
+        Method for running this component. It can be an Actor, Service or ServiceGroup.
+        It has to be defined in the child classes,
+        which is done in each of the default PipelineComponents.
```
Comment (RLKRo) [inline chatsky/pipeline/pipeline/component.py:157]: Honestly, the entire doc could be shortened to `Run this component`.

File: chatsky/pipeline/pipeline/component.py [hunk 15]
```diff
@@ -134,9 +150,35 @@ async def run_extra_handler(self, stage: ExtraHandlerType, ctx: Context, pipelin
...
+        :param pipeline: The current pipeline.
+        :return: `None`
+        """
         raise NotImplementedError
 
+    @property
+    def computed_name(self) -> str:
+        """
+        Every derivative of `PipelineComponent` must define this property.
+        :return: `str`.
```
Comment (RLKRo) [inline chatsky/pipeline/pipeline/component.py:175]:
```suggestion
        Default name that is used if `self.name` is not defined.
        In case two components in a `ServiceGroup` have the same `computed_name`
        an incrementing number is appended to the name.
```
Also, link the entities in the doc to their doc entries. You can link to `self.name` with 
```
:py:attr:`~.PipelineComponent.name`
```
The breakdown of the syntax:

- `~` is used to leave only the last component displayed (in markdown terms it's `[name](link)` instead of `[PipelineComponent.name](link)`). TBH this isn't necessary here; mostly used if a non-local object is referenced.
- `.` is used to reference a local object (`.Service` tells to find a `Service` entity in the current file; otherwise have to specify full path `chatsky.pipeline.service.service.Service`)
- `:py:attr:` specifies the type of the object referenced. Available types can be found here: https://www.sphinx-doc.org/en/master/usage/domains/python.html#cross-referencing-python-objects

File: chatsky/pipeline/pipeline/pipeline.py [hunk 16]
```diff
@@ -43,66 +44,95 @@ class Pipeline(BaseModel, extra="forbid", arbitrary_types_allowed=True):
...
+    This flag determines whether or not the functions
+    defined in the ``PRE_RESPONSE_PROCESSING`` and ``PRE_TRANSITIONS_PROCESSING`` sections
+    of the script should be parallelized over respective groups.
+    """
     _clean_turn_cache: Optional[bool]
 
-    def _create_actor(self) -> Actor:
+    @computed_field
+    @cached_property
+    def actor(self) -> Actor:
```
Comment (RLKRo) [inline chatsky/pipeline/pipeline/pipeline.py:135]: Could you add tests that these properties are not changed if any field of pipeline is changed?
```python
actor = id(pipeline.actor)
pipeline.start_label = other
assert actor == id(pipeline.actor)
```

File: chatsky/pipeline/pipeline/utils.py [hunk 17]
```diff
@@ -32,20 +29,10 @@ def rename_component_incrementing(component: PipelineComponent, collisions: List
...
-            base_name = component.handler.__name__
-        else:
-            base_name = component.handler.__class__.__name__
-    elif isinstance(component, ServiceGroup):
-        base_name = "service_group"
+    if isinstance(component, PipelineComponent):
+        base_name = component.computed_name
     else:
-        # This should never be triggered. (All PipelineComponent derivatives are handled above)
         base_name = "noname_service"
```
Comment (RLKRo) [inline chatsky/pipeline/pipeline/utils.py:35]: Now this definitely should never trigger.
Can either remove this if-else block or raise here.

Personally, I prefer removing the block:
- type of `component` can be validated with `validate_call`.
- it will raise if `component` does not have `computed_name`
- the block increases the number of lines

File: chatsky/pipeline/service/extra.py [hunk 18]
```diff
@@ -35,29 +35,39 @@ class ComponentExtraHandler(BaseModel, extra="forbid", arbitrary_types_allowed=T
...
+    """
+    Requested asynchronous property.
+    """
 
     @model_validator(mode="before")
     @classmethod
-    # Here Script class has "@validate_call". Is it needed here?
     def functions_constructor(cls, data: Any):
-        if not isinstance(data, dict):
+        if isinstance(data, (list, Callable)):
```
Comment (RLKRo) [inline chatsky/pipeline/service/extra.py:62]: This works but I think it's better to use the built-in `callable` instead of `isinstance(, Callable)`.

File: chatsky/pipeline/service/extra.py [hunk 19]
```diff
@@ -35,29 +35,39 @@ class ComponentExtraHandler(BaseModel, extra="forbid", arbitrary_types_allowed=T
...
             result = {"functions": data}
-        else:
+        elif isinstance(data, dict):
             result = data.copy()
+        else:
+            raise ValueError(
+                "Extra Handler can only be initialized from a Dict,"
+                " a Callable or a list of Callables. Wrong inputs received."
+            )
```
Comment (RLKRo) [inline chatsky/pipeline/service/extra.py:70]: No, `data` can be anything (e.g. it can be an instance of `ExtraHandler`).
Just do `return data` and pydantic will handle any type mismatches on its own.

So the code would be:
```python
if isinstance(data, list):
  return {"functions": data}
if callable(data):
  return {"functions": [data]}
return data
```

Same with some other validators.

File: chatsky/pipeline/service/group.py [hunk 20]
```diff
@@ -61,26 +52,41 @@ class ServiceGroup(PipelineComponent, extra="forbid", arbitrary_types_allowed=Tr
...
+    A `ServiceGroup` object, that will be added to the group.
+    """
+    # Inherited fields repeated. Don't delete these, they're needed for documentation!
+    before_handler: BeforeHandler = Field(default_factory=BeforeHandler)
+    after_handler: AfterHandler = Field(default_factory=AfterHandler)
+    timeout: Optional[float] = None
+    requested_async_flag: Optional[bool] = None
+    start_condition: StartConditionCheckerFunction = Field(default=always_start_condition)
+    name: Optional[str] = None
+    path: Optional[str] = None
```
Comment (RLKRo) [inline chatsky/pipeline/service/group.py:65]: Did `inherited-members` not work as expected?

File: chatsky/pipeline/service/group.py [hunk 21]
```diff
@@ -61,26 +52,41 @@ class ServiceGroup(PipelineComponent, extra="forbid", arbitrary_types_allowed=Tr
...
     @model_validator(mode="after")
-    def calculate_async_flag(self):
+    def __calculate_async_flag(self):
         self.calculated_async_flag = all([service.asynchronous for service in self.components])
         return self
 
-    async def run_component(self, ctx: Context, pipeline: Pipeline) -> None:
+    async def run_component(self, ctx: Context, pipeline: Pipeline) -> Optional[ComponentExecutionState]:
         """
         Method for running this service group. Catches runtime exceptions and logs them.
```
Comment (RLKRo) [inline chatsky/pipeline/service/group.py:91]: Do you need to copy the docstring?
I think the docstring should be inherited from `PipelineComponent` in the docs.

File: chatsky/pipeline/service/service.py [hunk 22]
```diff
@@ -94,6 +93,16 @@ async def run_component(self, ctx: Context, pipeline: Pipeline) -> None:
         else:
             raise Exception(f"Too many parameters required for service '{self.name}' handler: {handler_params}!")
 
+    @property
+    def computed_name(self) -> str:
+        if callable(self.handler):
```
Comment (RLKRo) [inline chatsky/pipeline/service/service.py:98]: I think this is guaranteed by the type annotation and `BaseModel` inheritance.
Try to get a 100% coverage on tests for pipeline.
We'd like to avoid unnecessary lines that are never reached (and cover others).

File: chatsky/pipeline/service/group.py (hunk 20 above)
Comment (ZergLev) [reply chatsky/pipeline/service/group.py:65]: Yes, it also inherits all methods of parent class, which are not needed, If I'm not wrong. Also, all inherited fields and methods are added at the end of the previous doc file, you could say. Meaning it didn't solve the issue at all and made it worse.
Using 'member-order': "groupwise" doesn't really help, because then methods go before fields for some reason. Though, within groups members are alphabetical, removing the distinction between differing classes. If the order can be tweaked, it would be fixed, but then `Actor` would look bad, because it doesn't need to inherit all of the fields in the docs. And inheriting all of the methods also seems unnecessary.

I think this is an okay solution for now. Either docs look good, or code looks good. This code is lacking in code-style, but the docs become broken/unreadable otherwise.
Well, maybe "broken docs" is just "not finished", really. I think there were extensions for Sphinx with Pydantic? Or extensions for sphinx-autodoc, those exist. This could take a long time, though.

File: chatsky/pipeline/pipeline/component.py (hunk 13 above)
Comment (ZergLev) [reply chatsky/pipeline/pipeline/component.py:73]: You mean that StartConditionCheckerFunction here should be a link, and the same for all entities in the docs? Or are there other missing links?

File: chatsky/pipeline/service/group.py (hunk 21 above)
Comment (ZergLev) [reply chatsky/pipeline/service/group.py:91]: True, I see ComponentExecutionState being described here, which is the only difference from PipelineComponent's docstring, but it's fairly obvious, could remove this.

File: chatsky/pipeline/pipeline/component.py (hunk 13 above)
Comment (RLKRo) [reply chatsky/pipeline/pipeline/component.py:73]: image
I think the field type annotation should already have a link to the `StartConditionCheckerFunction` (like in the screenshot with links to attachments).
If it is so, it's better to replace "StartConditionCheckerFunction" with "Function" in the docstring.

Otherwise "StartConditionCheckerFunction" should be turned into a link with `:py:data:`.

The issue I have with this line is that this is a reference to a library object yet it is in plain text so it looks like a regular word in the docs.

File: chatsky/pipeline/service/group.py (hunk 21 above)
Comment (RLKRo) [reply chatsky/pipeline/service/group.py:91]: Yeah, ok. If the component docstring is going to be as short as I suggested in the other comment then every concrete component should describe what it is doing during `run_component`.

File: chatsky/pipeline/service/extra.py (hunk 19 above)
Comment (ZergLev) [reply chatsky/pipeline/service/extra.py]: Fixed this, but my code can't look exactly like this, because `dict` has to be additionally validated - "functions" have to be cast to a list, since the passed `dict` could be anything.

File: chatsky/pipeline/service/service.py (hunk 22 above)
Comment (ZergLev) [reply chatsky/pipeline/service/service.py]: I saw that TYPE_CHECKING is often/always marked as uncovered code. The internet says adding "#pragma: no cover" to them should fix that, cause it's not supposed to be covered. Is there a reason we shouldn't do that?

**Review by RLKRo — APPROVED** [review]
